stat/roc_example_test.go

```diff
@@ -125,3 +125,126 @@ func ExampleROC_aUC() {
 	// false positive rate: [0 0.5 0.5 1 1]
 	// auc: 0.25
 }
+
+func ExampleROC_aUC_weighted() {
+	y := []float64{0.1, 0.35, 0.4, 0.8}
+	classes := []bool{true, false, true, false}
+	weights := []float64{1, 2, 2, 1}
+
+	tpr, fpr, _ := stat.ROC(nil, y, classes, weights)
+
+	// Compute Area Under Curve.
+	auc := integrate.Trapezoidal(fpr, tpr)
+	fmt.Printf("auc: %f\n", auc)
+
+	// Output:
+	// auc: 0.444444
+
+}
+
+func ExampleTOC() {
+	classes := []bool{
+		true, false, true, true, false, true,
+		false, true, true, false, false, false,
+		false, false, false, false, true, true,
+		true, true, true, true, true, true,
+		true, true, true, true, true, true,
+	}
+
+	min, ntp, max := stat.TOC(classes, nil)
+	fmt.Printf("minimum bound: %v\n", min)
+	fmt.Printf("TOC:           %v\n", ntp)
+	fmt.Printf("maximum bound: %v\n", max)
+
+	// Output:
+	// minimum bound: [0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 2 3 4 5 6 7 8 9 10]
+	// TOC:           [0 0 1 1 1 2 2 3 3 3 4 5 6 7 8 9 10 10 10 10 10 10 10 10 10 10 10 10 10 10 10]
+	// maximum bound: [0 1 2 3 4 5 6 7 8 9 10 10 10 10 10 10 10 10 10 10 10 10 10 10 10 10 10 10 10 10 10]
+}
+
+func ExampleTOC_unsorted() {
+	y := []float64{8, 7.5, 6, 5, 3, 0}
+	classes := []bool{true, true, true, false, true, false}
+	weights := []float64{2, 2, 3, 6, 1, 4}
+
+	stat.SortWeightedLabeled(y, classes, weights)
+
+	min, ntp, max := stat.TOC(classes, weights)
+	fmt.Printf("minimum bound: %v\n", min)
+	fmt.Printf("TOC:           %v\n", ntp)
+	fmt.Printf("maximum bound: %v\n", max)
+
+	// Output:
+	// minimum bound: [0 0 0 3 6 8 10]
+	// TOC:           [0 4 4 10 10 10 10]
```

Comment:
So I'm still completely confuddled by the parity change you made, and the correspondence (or seeming contradiction) to the docstrings, etc. My understanding was you wanted consistency with sklearn i.e. that values above or equal to the cutoff/ threshold/ value of the ranking variable are assigned class true. If that is the case, the TOC for this example should be [0 2 4 7 7 8 8].

Reply:
The issue comes down to; 1. the change asked for by #316 was to make ROC here behave the same way that sklearn's ROC behaves, and 2. that the AUC for a ROC curve will be the same as the AUC for a TOC parallelogram. If 1. was the correct thing to do then TOC has to behave this way AFAICS. Maybe it wasn't, but if that's the case, then that should be discussed and fixed.

Reply:
Yeah, maybe it's your point 2. that I don't understand then, because I still maintain the trapezoidal rule is an incorrect way to calculate the AUC for both ROC and a TOC because it incorrectly interpolates between points linearly, when this interpolation should actually be done in a stepwise manner (going left to right on the plot, horizontal and then vertical in the case of a ROC, horizontal and then parallel to the max bound in the case of a TOC). The trapezoidal rule calculates neither of these correctly in general, although in the case of TOC, so long as all cutoffs are calculated (as per the current implementation), it will actually happen to calculate the correct thing. In the case of ROC it will not, even in the case when all cutoffs are calculated, although in the case that a smaller number of cutoffs are used instead it might make more sense to use the trapezoidal rule approximation, and maybe that was the reasoning for doing it that way? If we don't allow that, as in this implementation of TOC, then maybe we should also go back to calculating the AUC correctly, but I can see why that should be dealt with in a seperate issue.

Reply:
Why do you think that it won't calculate correctly when the cutoffs are calculated? In that case, all (x,y) pairs are ends of straight lines along an ROC/TOC curve. If it's wrong, it's another issue though, as you say.

Reply:
Oh wait, you're completely correct and I'm not. This is actually completely fine. I was thinking about a completely different implementation of ROC calculations in which only vertical (TPR) changes are tracked, so the output is condensed... we don't do that here at all, so it's all good. Sorry for the confusion.

Reply:
:tada:

So the only remaining things is the API redesign to consider.

Reply:
Yep.